--- a/drivers/clocksource/Kconfig
+++ b/drivers/clocksource/Kconfig
@@ -223,4 +223,14 @@ config CLKSRC_VERSATILE
 	  ARM Versatile, RealView and Versatile Express reference
 	  platforms.
 
+config JZ47XX_TCU
+	bool
+
+config JZ4780_TCU
+	def_bool y if MACH_JZ4780
+	select JZ47XX_TCU
+	select CLKSRC_OF if OF
+	help
+	  Support for the timer/counter unit of the Ingenic jz4780 SoC.
+
 endmenu
--- a/drivers/clocksource/Makefile
+++ b/drivers/clocksource/Makefile
@@ -39,6 +39,8 @@ obj-$(CONFIG_FSL_FTM_TIMER)	+= fsl_ftm_t
 obj-$(CONFIG_VF_PIT_TIMER)	+= vf_pit_timer.o
 obj-$(CONFIG_CLKSRC_QCOM)	+= qcom-timer.o
 obj-$(CONFIG_MTK_TIMER)		+= mtk_timer.o
+obj-$(CONFIG_JZ47XX_TCU)	+= jz47xx-tcu.o
+obj-$(CONFIG_JZ4780_TCU)	+= jz4780-tcu.o
 
 obj-$(CONFIG_ARM_ARCH_TIMER)		+= arm_arch_timer.o
 obj-$(CONFIG_ARM_GLOBAL_TIMER)		+= arm_global_timer.o
