--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -1127,10 +1127,10 @@ config INDYDOG
 
 config JZ4740_WDT
 	tristate "Ingenic jz4740 SoC hardware watchdog"
-	depends on MACH_JZ4740
+	depends on MACH_JZ4740 || MACH_JZ4780
 	select WATCHDOG_CORE
 	help
-	  Hardware driver for the built-in watchdog timer on Ingenic jz4740 SoCs.
+	  Hardware driver for the built-in watchdog timer on Ingenic jz47xx SoCs.
 
 config WDT_MTX1
 	tristate "MTX-1 Hardware Watchdog"
